ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -33,119 +34,170 @@ import           Language.LSP.Server            (ProgressAmount (..),
-    = KickStarted
-    | KickCompleted
+  = ProgressNewStarted
+  | ProgressCompleted
+  | ProgressStarted
 
-data ProgressReporting  = ProgressReporting
-  { progressUpdate :: ProgressEvent -> IO ()
-  , inProgress     :: forall a. NormalizedFilePath -> Action a -> Action a
-  , progressStop   :: IO ()
+data ProgressReporting m = ProgressReporting
+  { progressUpdate :: ProgressEvent -> m (),
+    inProgress     :: forall a. NormalizedFilePath -> m a -> m a,
+    progressStop   :: IO ()
   }
 
-noProgressReporting :: IO ProgressReporting
-noProgressReporting = return $ ProgressReporting
-  { progressUpdate = const $ pure ()
-  , inProgress = const id
-  , progressStop   = pure ()
-  }
+noProgressReporting :: (MonadUnliftIO m) => IO (ProgressReporting m)
+noProgressReporting =
+  return $
+    ProgressReporting
+      { progressUpdate = const $ pure (),
+        inProgress = const id,
+        progressStop = pure ()
+      }
 
 -- | State used in 'delayedProgressReporting'
 data State
-    = NotStarted
-    | Stopped
-    | Running (Async ())
+  = NotStarted
+  | Stopped
+  | Running (Async ())
 
 -- | State transitions used in 'delayedProgressReporting'
 data Transition = Event ProgressEvent | StopProgress
 
 updateState :: IO () -> Transition -> State -> IO State
-updateState _      _                    Stopped       = pure Stopped
-updateState start (Event KickStarted)   NotStarted    = Running <$> async start
-updateState start (Event KickStarted)   (Running job) = cancel job >> Running <$> async start
-updateState _     (Event KickCompleted) (Running job) = cancel job $> NotStarted
-updateState _     (Event KickCompleted) st            = pure st
-updateState _     StopProgress          (Running job) = cancel job $> Stopped
-updateState _     StopProgress          st            = pure st
+updateState _ _ Stopped = pure Stopped
+updateState start (Event ProgressNewStarted) NotStarted = Running <$> async start
+updateState start (Event ProgressNewStarted) (Running job) = cancel job >> Running <$> async start
+updateState start (Event ProgressStarted) NotStarted = Running <$> async start
+updateState _ (Event ProgressStarted) (Running job) = return (Running job)
+updateState _ (Event ProgressCompleted) (Running job) = cancel job $> NotStarted
+updateState _ (Event ProgressCompleted) st = pure st
+updateState _ StopProgress (Running job) = cancel job $> Stopped
+updateState _ StopProgress st = pure st
 
 -- | Data structure to track progress across the project
-data InProgressState = InProgressState
-    { todoVar    :: TVar Int  -- ^ Number of files to do
-    , doneVar    :: TVar Int  -- ^ Number of files done
-    , currentVar :: STM.Map NormalizedFilePath Int
-    }
+data InProgressState
+  = InProgressState
+      { -- | Number of files to do
+        todoVar    :: TVar Int,
+        -- | Number of files done
+        doneVar    :: TVar Int,
+        currentVar :: STM.Map NormalizedFilePath Int
+      }
+  | InProgressStateOutSide
```

Comment:
I still find this split a bit weird, but we've spent enough time on this!

Reply:
yes, especially we have `inProgress` that is useless `InProgressStateOutSide`.
Maybe in some following ups, we can come up with alternative design such as using gadt to eliminate the oddness